### ISSUE TITLE
CONTRIBUTION.md should be updated to specify what code style should be applied to PRs

### ISSUE THREAD

**Issue by riccardobl** [jMonkeyEngine/jmonkeyengine#1098]
Following : https://github.com/jMonkeyEngine/jmonkeyengine/issues/1056
It seems that the common agreement is that we should use netbeans style, if that is the case, we should write that down on CONTRIBUTION.md and provide settings for at least the most common ides.

This link contains the configurations for eclipse and other compatible editors (eg vscode) https://we.tl/t-jcnutcbB4U

Since these days, many people use idea, someone should also check if the default idea style is ok for us or if the configuration provided above can be applied (with extensions or whatever) or if there is another way to define and export the style.

**Comment by louhy** [comment]
Some other things to consider adding there if adopting standards, with reasoning below. 

1. Default NetBeans defaults to an **80 character right margin**. I have mixed feelings about this. Many existing comments will be wrapped as a result (creating more Git chaos). It seems outdated with today's larger monitors. But, it may allow you to snap your IDE to one half of the monitor without cutting off code. And I guess it helps when we're forced to use craptops. :stuck_out_tongue: 

2. Should there be (or is there) a guideline on commented out code? Git has the history for things removed, and "WIP" stuff should be documented. If something is being removed but there's doubt that it's correct to do so, at least _add a comment explaining why it was disabled_. 
**TLDR: If you check in (new) commented out code, explain why.** Otherwise others are wasting time wondering "why is this here?", and it just detracts from code readability/quality. Experimental code under very heavy development working a tricky problem can be an exception here... 

3. **Remove Unused Imports** - NB and Eclipse can do it with a single key or "On Save" (for NB). I don't know why this isn't enabled by default when it's usually also a single keypress to re-add them if needed. 

4. Here's where I really get annoying... **Try to address warnings/suggestions as much as possible.** Your IDE will often offer suggestions to clean up warnings or improve code with a single click. Sometimes it's not always possible, and not all IDEs will agree on these, but warnings are there for a reason so they shouldn't be ignored totally. If you ignore 100 "not important" warnings, you'll miss the one that would have helped you catch a mistake... which is the whole point of warnings. Also, it annoys the heck out of people who want to use the right-gutter to navigate a file, find references, and so on when there's dozens of yellow flag warnings cluttering it up. I go as far as using suppression annotations if necessary (rarely), ex. `@SuppressWarnings("unchecked")`... should be as narrow as possible, don't suppress all warnings. 

I realize many of these may be an un-winnable battle to fight, but you don't know if you don't try. 

Feel free to disagree if these are unpopular; won't hurt my feelings. lol (Actually, if anyone has strong arguments against any of these, I'd rather hear them!)

**Comment by riccardobl** [comment]
I think we shouldn't make contributions bureaucratic to the point of regulating comments...
and 3 and 4 are based on ide specific features.
Also remember that PRs still need to be approved, so if someone is doing something weird, like adding imports that are not used or commenting half of the code, he can always be asked to improve it.

I'm fine with point 1.

**Comment by pspeed42** [comment]
Yeah, personally, I set my command prompt, editor, etc. widths to 110 a loooong time ago.  80 characters is a hold over from when that's all displays had and we were editing in DOS full screen. :)

**Comment by louhy** [comment]
I agree (I hate bureaucracies), that's why I think I tried to phrase as "guidelines".

**Comment by louhy** [comment]
@riccardobl 3 - depends on IDE support, but I feel like all 3 of the big ones have support for it. Only one I'm not sure about is IJ. It's certainly less important than the others. 

4 - Code suggestions = IDE specific for sure, that's a fair point. Compiler warnings such as those suppressed via @SuppressWarnings, well I guess they're Java _version_ specific, but that's definitely coming from Java (and they're the important ones). 

I realize I'm in the minority on these for the most part (rare to see "clean" code), but I do feel it's fundamentally better to address warnings if possible. 

Whether these are worth writing up as contribution guidelines, dunno. PRs can always be cleaned up, but if they're good to start with, it's less work for the reviewer.

@pspeed42 Borland IDE days? I can't remember that far back...

**Comment by NetzwergX** [comment]
As I am working on a PR right now... has there been any consensus about a Code style yet? Is there a doc describing the preferred code style for new contributions? It would be really helpful. The class(es) I am working on have mis-matched code styles all over the place and I would really like to start cleaning that up when touching it.

**Comment by stephengold** [comment]
I believe there's consensus on spacing, indentation, and placement of brackets, at least. I'm unconvinced regarding line width.

If you wish to clean up formatting, I suggest you make that a separate PR from any other (substantial/functional) changes you wish to contribute.

**Comment by louhy** [comment]
Using the NetBeans formatter's defaults seemed agreeable to most and I think it's fairly close to the proposed Google style guide (in the other issue)? 

It's a decent starting point that I've been using, but IMO still an imperfect solution, even if everyone agreed on it. Not certain how big an issue this is, but I recall it mangling some JME comments when I let it loose to just format entire files. If we do any reformats of big files we need to check that it hasn't trashed anything. I found the NetBeans feature to "only format changed lines" helpful, of course then it might be a little inconsistent within the file. :weary: 

I doubt this will ever be fully addressed unless someone dares to take on reformatting the entire project all at once (which brings other issues).

**Comment by NetzwergX** [comment]
> If you wish to clean up formatting, I suggest you make that a separate PR from any other (substantial/functional) changes you wish to contribute.

I'd usually only clean up the code I am touching anyways.

**Comment by stephengold** [comment]
A draft proposal:
+ Our preferred style for Java source code is [Google style](https://google.github.io/styleguide/javaguide.html) with the following mods:
   1. Column limit of 110 instead of 100.
   2. Block indentation of 4 spaces instead of 2.
+ Any PR that adds new Java source files shall apply our preferred style to those files.
+ Any PR that has reformatting as it's primary purpose shall apply our preferred style to every file it modifies.
+ Any PR that modifies a pre-existing source file AND doesn't have reformatting as it's primary purpose shall either
  1. conform to the prevailing style of that file OR
  2. apply our standard style, *but only to the portions modified for the PR's primary purpose* 

The intent is:
+ Promote a gradual transition to a uniform style, one that works for most developers.
+ Don't mix reformatting with PRs for other purposes, to ease code review.
+ Define a preferred style that matches that produced by the Netbeans formatter, except for comments and column limit, without referring to Netbeans explicitly.

Any questions or reactions?

**Comment by stephengold** [comment]
As an experiment, I'm going to submit a reformatting PR in order to see what problems arise, if any.

**Comment by stephengold** [comment]
Since we specify 4 spaces for block indentation, we should specify "at least +8 spaces" for continuation lines.

**Comment by stephengold** [comment]
Another thing ... Google style requires exactly one blank line between the license and the `package` statement, if both are present. However, JMonkeyEngine's prevailing style has no blank line before the `package` statement, so I recommend omitting that requirement from our preferred style.

**Comment by stephengold** [comment]
Google style requires "some logical ordering" of class contents (3.4.2) but this is difficult to follow/enforce unless the content order is either (a) obvious or (b) documented. Perhaps we should omit this requirement from our preferred style.

**Comment by stephengold** [comment]
Google style apparently requires that "//" comments be indented to match the surrounding code, making no exception for commented-out code. In our codebase, it's typical for commented-out code to have "//" at the start of the line, not indented. Shall we make an exception to allow this?

Also, Google style requires that names of test classes end with "Test", whereas our prevailing style puts "Test" at the beginning of the name. Should we override this requirement?

**Comment by stephengold** [comment]
Applying Google style to existing code will sometimes require API changes. For instance renaming `IDList` to `IdList` per the section 5.3 definition of camel case. Do we want to create a policy around this?

**Comment by pspeed42** [comment]
I'm personally ok with exceptions for the other cases mentioned.  But regarding things like IDList, it would be good to put a plan in motion to eventually fix these.  Document them for now as being grandfathered in, create parallel properly-named ones in the cae of methods (and deprecate the old ones).

I don't know where IDList comes up exactly but it's a public or protected field then that's going to just be another unfortunate example of where that anti-idiom bites us.

**Comment by stephengold** [comment]
`IDList` is a public class in "com.jme3.render", but I don't know any reason why an app would use it.

**Comment by stephengold** [comment]
After spending some "quality time" with the Google Java Style Guide, I believe we should encourage PRs that improve style incrementally, instead of applying the entire set of guidelines at once. There are a lot of guidelines!

I've also come to realize that there's a distinction between "reformatting" and "style improvement".  A lot of the Guide has nothing to do with formatting, so I'll stop using those terms interchangeably.

Here's a revised draft proposal:

+ Our preferred style for Java source code is [Google style](https://google.github.io/styleguide/javaguide.html) with the following 7 modifications:
  1. No blank line before a `package` statement. (Section 3)
  2. Logical ordering of class contents is encouraged but not required. (Section 3.4.2) 
  3. Block indentation of +4 spaces instead of +2. (Section 4.2)
  4. Column limit of 110 instead of 100. (Section 4.4)
  5. Continuation line indentation of +8 spaces instead of +4. (Section 4.5.2)
  6. Commented-out code need not be indented at the same level as surrounding code. (Section 4.8.6.1)
  7. The names of test classes need not end in "Test". (Section 5.2.2)
+ Any PR that adds new Java source files shall apply our preferred style to those files.
+ Any PR that has style improvement as its primary purpose shall apply our preferred style, or specific aspect(s) thereof, to every file it modifies.
+ Any PR that modifies a pre-existing source file AND doesn't have style improvement as it's primary purpose shall either:
  1. conform to the prevailing style of that file OR
  2. apply our preferred style, but only to the portions modified for the PR's primary purpose.

Please closely examine PR #1653, which attempts to apply "preferred style" to 12 files in the `com.jme3.renderer` package.

+ Do you see any undesirable or unnecessary changes in PR 1653?
+ Do you see any residual style issues in those files that our style guide should address?

**Comment by tlf30** [comment]
Another one to look at for reference is Twitter'a style guide. My company adopted a modified version of it. It is short, simple, and effective.

**Comment by stephengold** [comment]
At first glance, the Twitter guide doesn't seem any shorter or simpler than the Google guide. Are there particular parts of it that you want us to incorporate?

**Comment by tlf30** [comment]
Ah, I guess it has gotten larger since I looked at it a couple years ago. 
Here is the one that my company uses, which was a modified version of twitters from several years ago.
https://github.com/tlf30/monkey-netty/blob/master/STYLEGUIDE.md

Nothing in particular without spending more time digging. Just figured it was a good reference. See the one I linked, it is simpler and more to my liking than a large complex one.

**Comment by stephengold** [comment]
One thing I'm tempted to adopt from your style guide is the "No trailing whitespace" rule. Does anyone have an opinion on this?

**Comment by tlf30** [comment]
That would be fine with me, trailing whitespace can be problematic.

**Comment by pspeed42** [comment]
OMG, yes.  I finally switched all of my editor configs to chop it and JME is a project where it creates unwanted changes.

(At day job it's fun to watch the white space wars between devs have configure with no trailing whitespace and those that don't.)

**Comment by stephengold** [comment]
I added the revised draft proposal plus "No trailing whitespace" to the document and submitted PR #1655.

**Comment by stephengold** [comment]
Google style 7.3.1 says that javadoc is optional for "simple, obvious" methods like getFoo(). However, OpenJDK 15 and later will generate "no comment" warnings for such public methods. In the interest of reducing the number of warnings, perhaps we shouldn't make exceptions for "simple, obvious" methods.

**Comment by tlf30** [comment]
I personally think all methods should have a javadoc, unless the method is an `@Override` in which case it should only have a javadoc if the methods behavior differs from the parent's method.

**Comment by pspeed42** [comment]
I would think there must be a way to turn off that warning.

If all methods are forced to have javadoc then you end up with a lot of really stupid javadoc... which in the end is worse than no javadoc at all.  (My favorite is the IDE generated garbage javadoc.)  Because at least it's easy to find missing javadoc versus bad javadoc.

```
/**
 *  Returns foo.
 */
```
Reminds me of the old "all code must have comments" stuff that led to (in the old days) stuff like this:
```
// Increment i
i++;
```
If you are going to have a "Returns foo." comment then it better also include what foo is, why it's used, whether it can be null, is there a default... whatever.  Else just leave it off.

**Comment by Sailsman63** [comment]
The "No Comment" errors would be coming from the `-Xdoclint` option, which is enabled by the build scripts *when building javadoc*. Suppressing it would mean adjusting options in the buildfiles.

**Comment by stephengold** [comment]
Regarding doclint, I believe it's possible to disable just this one warning:  https://stackoverflow.com/questions/39333574/use-maven-javadoc-with-reasonable-doclint-parameters

As far as whether to require javadoc (for all public and protected members, excluding deprecated or overrides, in public classes), I'm waffling. (BTW, thanks Paul for providing some of the rationale behind your opinion!)

90-99% of the time, when the developer omits javadoc from such methods, it's due to some combination of
+ laziness (not wanting to think about or type javadoc) and
+ self-centered thinking (it's obvious to me what the method does, so it'll be obvious to everyone).

In my personal projects, I try to come up with something non-obvious to say about every method I write. That effort often leads me to:
+ realize that the method (or its formal arguments) could be named better, or 
+ question whether the method is necessary, or 
+ decide to improve the implementation (add a range check, for instance).

I think the current "simple, obvious" loophole will lead to fewer such realizations/questions/decisions. But perhaps benefits don't justify the extra effort.

"Obvious" is a subjective opinion. Keeping javadoc optional for obvious methods seems likely to cause heated discussion between developers and reviewers about what's obvious. But perhaps such such discussion is a good thing.

It's tricky (for me) to think about these matters in the abstract. It's easy to find examples of boilerplate javadoc that doesn't convey any useful information. What's hard (for me) is thinking of a method (in a public class, public or protected, excluding deprecated or overrides) that *shouldn't* have javadoc. Would someone who favors the "simple, obvious" loophole cite an example of such a method, from practical open-source code, for us to discuss?

**Comment by pspeed42** [comment]
If javadoc is always required then it means two things, either:
1. we are ok with lots of incomplete placeholder javadoc
2. we are ok with having no code at all for some particular thing

Unjavadoc'ed code can eventually be documented.  Having no code at all can never be documented.

**Comment by stephengold** [comment]
Mhh. I'd go with less crappy code to maintain, even if it means fewer features in our Engine. So much code has been added without serious consideration for support and/or maintenance.

But if there's no such thing as a "simple, obvious" class member, then the loophole in question is never applicable.

**Comment by pspeed42** [comment]
Properly documenting every method and protected/public field can easily take twice as long as writing the code itself.  I'd take good code with minimal but good javadoc over no code... and especially over code that has 100% javadoc but 90% of it is useless "Returns foo" style stuff.  I can read code.  I can't read a blank page.  And crappy javadoc is harder to find than missing javadoc.

And until all of JME adheres to said requirement (hah!) then it doesn't make sense to impose it on all new code where some getters are obvious, etc...  Submissions should be gated on whether there is enough documentation/good-coding-practice for the rest of the contributors to understand it... rather than some arbitrary 100% completion metric.

And anyway, until JME complies then if we can turn off that ludicrous warning in the javadoc that is easily determined in a multitude of other ways then at least we'll see other real issues in the javadoc.

**Comment by Sailsman63** [comment]
While acknowledging that "Trivial, obvious" is very subjective, I think that one can set some (low) objective bars.

At the very least, if a getter or setter does nothing other than literally pass through to or from a field, what would a Javadoc comment *say*? `Returns the foo field with no other processing`?

**Comment by stephengold** [comment]
It's difficult to craft good documentation for a method that doesn't exist and has no context. That's why I requested an example. But OK, I'll use my imagination...

The javadoc could document the units of "foo" or how the returned quality is defined/measured. If `foo` is an angle, is it in degrees or radians? Measured counterclockwise? If it's a ratio, what are the numerator and denominator? If it's a bit position, is it bigendian or littleendian?

The javadoc could talk about the expected range of return values and/or the default value. Can the returned value be null, empty, NaN, infinite, or negative? If "foo" is a vector/quaternion/matrix/color, the javadoc should say whether the returned object is an alias or a copy. If it's a `Joint` or `Spatial`, can it be a leaf and/or a root? If `foo` is a collection, does `getFoo()` return a modifiable instance? If it's a location/offset/direction/rotation, which system (world or local) is it measured in? Does `getFoo()` use special values (such as `null` or -1) to indicate an error or abnormal condition?

If a `public` or `protected` method is intended solely for the library's internal use, the javadoc should say so. If the method was recently added, it might have `@since` javadoc. If the method exists solely to comply with some standard, or for compatibility with another library, then say so. If `getFoo()` is more efficient (or less accurate) that `grindOutFooTheHardWay()`, then say so.

It might be helpful to link to the corresponding setter, if one exists, or the `foo` field itself, if it's not private. If "foo" is known by other names, those should be mentioned.  Is "foo" an abbreviation for "Fancy Octal Operator"? Or is it short for "fubar"?

If the returned value isn't valid until `update()` has been invoked and/or unless `isFooValid()` returns `true`, those caveats should be mentioned.

**Comment by stephengold** [comment]
> Properly documenting every method and protected/public field can easily take twice as long as writing the code itself.

"Twice as long" doesn't scare me. Many good software practices involve extra time and effort: writing requirements and design docs, creating and running tests, stepping through in a debugger, reviewing PRs, and responding to review comments. Often the benefits outweigh the costs, especially when you include support and maintenance costs later on.

> And until all of JME adheres to said requirement (hah!) then it doesn't make sense to impose it on all new code

In proposing style guidelines, I've tried to take prevailing practices into account. But so much of the prevailing practice is intolerably bad and/or hard to explain that any attempt at guidelines will inevitably raise the bar.

**Comment by Sailsman63** [comment]
I'm coming around to @stephengold's thought here. Those are all things that would be worth documenting, and I'm starting to think that if there is *nothing* worth documenting, then perhaps you're using the wrong tool.

If the only purpose of a `public` method is to provide an accessor to a `protected/private` field, why not just open up the field visibility and use it directly? (See `com.jme3.renderer.Camera.getFrustrum<direction>()` methods, which provided the spark for this particular bit of discussion)

**Comment by stephengold** [comment]
I'm still ambivalent about the loophole (7.3.1) for javadoc of "simple, obvious" members. On the one hand, Paul has argued strenuously that the loophole is needed to avoid burdening developers with unnecessary work. On the other hand, Trevor thinks that every non-override method should have javadoc. Both viewpoints have merit (and emojis). Neither seems likely to convince the other.

I'm still not convinced the loophole applies often in practice.

With that uncertainty in mind, I I'd like to move forward with #1655. I propose to leave the loophole in. If it causes too much friction, future team members can amend CONTRIBUTION.md to override section 7.3.1 .
